snuba/datasets/query_plan_delegator.py

```diff
@@ -0,0 +1,159 @@
+)
+from snuba.datasets.plans.translator.query import identity_translate
+from snuba.request import Request
+from snuba.request.request_settings import RequestSettings
+from snuba.util import with_span
+from snuba.web import QueryResult
+
+
+BuilderId = str
+QueryResults = List[Tuple[BuilderId, QueryResult]]
+CallbackFunc = Callable[[QueryResults], None]
+
+
+class QueryPlanDelegator(ClickhouseQueryPlanBuilder):
+    """
+    QueryPlanDelegator is a query plan builder that delegates a request to one or
+    more other query plan builders.
+
+    It can be used as a way to switch between different query plan builders depending
+    on the query received and the selector function provided. It can also be used to
+    run a single query against multiple query plans annd their associated storages
+    in order to test and compare the results and performance of various query
+    plan builders before rolling them out to users.
+
+    The query plans that are built and executed on a particular query are determined
+    by the selector function which returns a sequence of query builder IDs. The
+    first item in the list corresponds to the query builder that will be used to
+    ultimately return a result. Hence the selector function must return
+    a list with at least one value in it.
+
+    Subsequent (if any) query plan builders that are also returned by the selector
+    function will be executed in separate threads. The results of these other queries
+    are provided to the callback function if one is provided.
+
+    TODO: Add metrics
+    """
+
+    def __init__(
+        self,
+        query_plan_builders: Mapping[BuilderId, ClickhouseQueryPlanBuilder],
+        selector_func: Callable[[Query], List[BuilderId]],
+        callback_func: Optional[CallbackFunc] = None,
+    ) -> None:
+        self.__query_plan_builders = query_plan_builders
+        self.__selector_func = selector_func
+        self.__callback_func = callback_func
+
+    def build_plan(self, request: Request) -> ClickhouseQueryPlan:
+        return ClickhouseQueryPlan(
+            query=identity_translate(request.query),
+            plan_processors=[],
+            execution_strategy=ThreadedQueryPlanExecutionStrategy(
+                request=request,
+                query_plan_builders=self.__query_plan_builders,
+                selector_func=self.__selector_func,
+                callback_func=self.__callback_func,
+            ),
+        )
+
+
+class ThreadedQueryPlanExecutionStrategy(QueryPlanExecutionStrategy):
+    def __init__(
+        self,
+        request: Request,
+        query_plan_builders: Mapping[BuilderId, ClickhouseQueryPlanBuilder],
+        selector_func: Callable[[Query], List[BuilderId]],
+        callback_func: Optional[CallbackFunc] = None,
+    ) -> None:
+        self.__request = request
+        self.__query_plan_builders = query_plan_builders
+        self.__selector_func = selector_func
+        self.__callback_func = callback_func
```

Comment:
I don't think this approach can work for joined queries and subqueries. In that case we need to be able to build a full execution plan for each subquery (we will likely discard the execution strategies), then apply query processors and translation to each query individually and reassemble the queries in a new composite query that we would pass to a composite execution strategy.
If the query translation and processing logic moves into the execution strategy, the process above will not work. Moreover it would break several invariannts of our architecture like fact that execution strategies only tweak the query for optimizations, and do not impact correctness. 

I assume you followed this path since, otherwise, you would not be able to instruct the query plan builder to build multiple different plans that then you can execute. 
In such case, it means we are missing some abstractions here that we should add.
The current query plan and query plan builder is meant to provide the building blocks that something external (web/query.py)  would use to execute a query. IT does not provide any logic to fully run the query.

What about adding a QueryPipeline abstraction that takes in a query and returns a result. We can keep it small and cover only the physical part of the query. The entity could provide a PipelineBuilder. that Produces a Pipeline that can run (when provided an executor). 

Under normal circumstances, web/query would use the Pipeline builder provided by the entity in the same way it uses the QueryPlanBuilder, and then call pipeline.query(executor) to get the result.

In your case the pipelline builder would generate a pipelinedelegate that encapsulates two pipelines executed in parallel. 

In the case of a join we would have a composite pipeline that executes query processing for multiple subqueries and then executes the composite query.

This way, instead of fitting the query processing into an existing components that is not supposed to do it, we build something on top of the current processing that coordinates the execution and can execute multiple queries in parallel without changing the semantics of the existing components.

How does it sound?

(also I would introduce the pipeline in its own pr since it will be verbose)